packages/amplify-graphql-auth-transformer/src/__tests__/relational-field-redaction.test.ts

```diff
@@ -0,0 +1,263 @@
+    } else {
+      expect(out.resolvers['Primary.relatedMany.auth.1.req.vtl']).not.toContain(SUBSCRIPTION_PROTECTION);
+      expect(out.resolvers['Primary.relatedOne.auth.1.req.vtl']).not.toContain(SUBSCRIPTION_PROTECTION);
+      expect(out.resolvers['RelatedMany.primary.auth.1.req.vtl']).not.toContain(SUBSCRIPTION_PROTECTION);
+      expect(out.resolvers['RelatedOne.primary.auth.1.req.vtl']).not.toContain(SUBSCRIPTION_PROTECTION);
+    }
+  };
+  return expectation;
+};
+
+export type TestTableRow = [string, string[], string[], boolean];
+
+/**
+ * Primary auth rules - Related auth rules - Redact relational field
+ */
+const testCases: [string[], string[], boolean][] = [
+  // Cases for field redaction
+  [['public apiKey', 'private userPools'], ['private userPools'], true],
+  [['public apiKey', 'owner userPools'], ['public apiKey', 'owner userPools'], true],
+  [['custom'], ['custom'], true],
+  [['dynamic groups userPools'], ['dynamic groups userPools'], true],
+  [['private userPools'], ['owner userPools'], true],
+  [['static groups userPools multiple'], ['static groups userPools single'], true],
+  [['private oidc', 'dynamic groups userPools'], ['private oidc', 'dynamic groups userPools'], true],
+  // Cases for non field redaction
+  [['private userPools', 'owner userPools'], ['private userPools'], false],
+  [['private oidc', 'dynamic groups oidc'], ['private oidc', 'dynamic groups oidc'], false],
+  [['private userPools'], ['private userPools'], false],
+  [['private oidc', 'static groups oidc single'], ['private oidc', 'static groups oidc single'], false],
+  [['owner oidc'], ['owner userPools'], false],
+  [['owner userPools'], ['owner userPools claim'], false],
+  [['public iam'], ['private iam'], false],
+];
+
+describe('Relational field redaction tests', () => {
+  describe('DDB datasources - field relation', () => {
+    const expectation = makeTransformationExpectation(ddbDataSourceStrategies, fieldSchemaTemplate);
+    const testTable: TestTableRow[] = [];
+    for (const strategyName of Object.keys(ddbDataSourceStrategies)) {
+      testCases.forEach((testCase) => {
+        testTable.push([strategyName, ...testCase] as TestTableRow);
+      });
+    }
+    test.each(testTable)(`%s - Primary %s - Related %s should redact relational field - %s`, expectation);
+  });
+  describe('DDB datasources - reference relation', () => {
+    const expectation = makeTransformationExpectation(ddbDataSourceStrategies, referenceSchemaTempalte);
+    const testTable: TestTableRow[] = [];
+    for (const strategyName of Object.keys(ddbDataSourceStrategies)) {
+      testCases.forEach((testCase) => {
+        testTable.push([strategyName, ...testCase] as TestTableRow);
+      });
+    }
+    test.each(testTable)('%s - Primary %s - Related %s should redact relational field - %s', expectation);
+  });
+  describe('RDS datasources - reference relation', () => {
```

Comment:
It "shouldn't" make a difference, but we should test heterogeneous relationships as well (DDB Primary, SQL Related; and SQL Primary, DDB Related)